File: 2020/s2.py
#!/bin/python3
import sys

# Recursion limit needs to be high
sys.setrecursionlimit(1_000_000)


rows = int(sys.stdin.readline())  # No. of rows
columns = int(sys.stdin.readline())  # No. of columns

"""Visited squares.
We only need one visited set because if the square has been visited we know
there's no way that square will take us to the end.
This way, we don't look through the same path multiple times.
"""
visited = set()

"""Squares to be stored in the following way:
{valueOfSquare: [productOfCoords...]}
"""
squares = {}

for i in range(rows):
    vals = list(
        map(int, sys.stdin.readline().split())
    )  # Split input by space and convert values to an integer

    for index, val in enumerate(vals):  # Add squares to the squares dictionary
        squareValue = (i + 1) * (index + 1)
        squares.setdefault(val, []).append(squareValue)


def solve(squareValue):
    # If the value of the square is 1, it can only go to (1, 1).
    # Therefore, we have reached the end.
    if squareValue == 1:
        return True

    for index in squares.setdefault(squareValue, []):
        if index not in visited:  # If square hasn't been visited
            visited.add(index)  # Add to visited
            if solve(index):  # Solve
                return True

    return False


print("yes" if solve(rows * columns) else "no")

File: 2020/s2.cpp
#include <bits/stdc++.h>
#include <iostream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using std::cin;
using std::cout;
using std::endl;
using std::pair;
using std::string;
using std::vector;

#define set std::unordered_set
#define map std::unordered_map

uint rows;    // No. of rows
uint columns; // No. of columns

/**
 * Visited squares.
 * We only need one visited set because if the square has been visited we know
 * there's no way that square will take us to the end.
 * This way, we don't look through the same path multiple times.
 */
set<uint> visited;

/**
 * Squares to be stored in the following way:
 * {valueOfSquare: [productOfCoords...]}
 */
map<uint, vector<uint>> squares;

/**
 * Splits str by spaces and returns a vector
 */
vector<string> split(string str) {
    vector<string> words;
    std::istringstream ss(str);
    string word;

    while (ss >> word) {
        words.push_back(word);
    }

    return words;
}

/**
 * DFS resurcive function to solve problem
 */
bool solve(uint squareValue) {
    // If the value of the square is 1, it can only go to(1, 1).
    // Therefore, we have reached the end.
    if (squareValue == 1) {
        return true;
    } else if (squares.find(squareValue) == squares.end()) { // If square with
                                                             // value doesn't
                                                             // exist, return
        return false;
    }

    for (auto& index : squares[squareValue]) {
        if (visited.find(index) ==
            visited.end()) {       // If square hasn't been visited
            visited.insert(index); // Add to visited
            if (solve(index)) {    // Solve
                return true;
            }
        }
    }

    return false;
}

int main() {
    cin >> rows >> columns;
    cin.ignore(); // Ignore newline

    for (uint i = 0; i < rows; i++) {
        string vals;
        std::getline(cin, vals);

        auto splitVals = split(vals); // Split input by space

        // Add squares to the squares map
        for (uint index = 0; index < splitVals.size(); index++) {
            uint val = std::stoi(splitVals[index]);
            uint squareValue = (i + 1) * (index + 1);

            if (squares.find(val) == squares.end()) {
                vector<uint> coordinateProduct = {squareValue};
                squares[val] = coordinateProduct;
            } else {
                squares[val].push_back(squareValue);
            }
        }
    }

    cout << (solve(rows * columns) ? "yes" : "no") << endl;
}
